File: ResolucaoProvas/2025.1/C/q2.c
#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>
#include <semaphore.h>
#include <unistd.h>
#include <time.h>

#define M 100
#define CONSUMER_N 8

int buffer[M];
int buffer_size = M;
int buffer_index = M - 1;
sem_t mutex;
sem_t producer;
sem_t consumer;

int getJob(pthread_t tid) {
    printf("Consumer %lu getting a job\n", (unsigned long)tid);
    int job = buffer[buffer_index];
    buffer_index--;
    return job;
}

void execute(int j) {
    printf("Executing task: %d\n", j);
    sleep(1);
}

void* compute(void* arg) {
    while(1) {
        sem_wait(&mutex);

        if(buffer_size == 0) {
            sem_post(&producer);
            sem_post(&mutex);
            sem_wait(&consumer);
            sem_wait(&mutex);
        }

        int j = getJob(pthread_self());
        buffer_size -= 1;
        sem_post(&mutex);

        execute(j);
    }
    return NULL;
}

void create_jobs(int M) {
    printf("Starting job creation\n");
    for(int i = 0; i < M; i++) {
        buffer[i] = i + 1;
    }
    buffer_index = M - 1;
    sleep(2);
    printf("Job creating done!\n");
}

void* manage(void* arg) {
    while(1) {
        sem_wait(&producer);
        create_jobs(M);
        buffer_size = M;
        sem_post(&consumer);
    }
    return NULL;
}

int main() {
    pthread_t consumerThreads[CONSUMER_N];
    pthread_t producerThread;

    sem_init(&mutex, 0, 1);
    sem_init(&producer, 0, 0);
    sem_init(&consumer, 0, 0);

    for(int i = 0; i < M; i++) {
        buffer[i] = i + 1;
    }

    pthread_create(&producerThread, NULL, manage, NULL);

    for(int i = 0; i < CONSUMER_N; i++) {
        pthread_create(&consumerThreads[i], NULL, compute, NULL);
    }

    pthread_join(producerThread, NULL);
    for(int i = 0; i < CONSUMER_N; i++) {
        pthread_join(consumerThreads[i], NULL);
    }

    sem_destroy(&mutex);
    sem_destroy(&producer);
    sem_destroy(&consumer);

    return 0;
}

File: ResolucaoProvas/2025.1/Prova1/Python/q2.py
import threading
import random
import time

M = 100
buffer_size = M
buffer = list(range(1,M+1))
mutex = threading.Semaphore(1)
producer = threading.Semaphore(0)
consumer = threading.Semaphore(0)

def compute():
    global buffer_size
    global producer
    global consumer
    while(True):
        mutex.acquire()
        
        if (buffer_size == 0):
            producer.release()
            consumer.acquire()
        
        j = getJob()
        buffer_size -= 1
        mutex.release()

        execute(j)

def getJob() -> int:
    global buffer
    print(f"{threading.current_thread().name} getting a job")
    return buffer.pop()

def execute(j: int):
    print(f"Executing task: {j}")
    time.sleep(1)

def manage():
    global producer
    global consumer
    global M
    global buffer_size
    while(True):
        producer.acquire()
        create_jobs(M)
        buffer_size = M
        consumer.release()

def create_jobs(M: int):
    global buffer
    print("Starting job creation")
    for i in range(1, M+1):
        buffer.append(i)
    time.sleep(2)
    print("Job creating done!")


if __name__ == "__main__":
    consumerN = 8
    consumerThreads = []
    all_threads = []
    producerThread = threading.Thread(target=manage, name="Producer")

    for i in range(consumerN):
        consumerThreads.append(threading.Thread(target=compute, name=f"Consumer {i + 1}"))
    
    all_threads = consumerThreads
    all_threads.append(producerThread)
    random.shuffle(all_threads)

    for thread in all_threads:
        thread.start()
    
    for thread in all_threads:
        thread.join()
